sandbox/cgroup/cgroup.go

```diff
@@ -260,6 +258,119 @@ func MockVersion(mockVersion int, mockErr error) (restore func()) {
+		// least two dots exist.
+		dot1 := strings.IndexRune(leaf, '.')
+		dot2 := strings.IndexRune(leaf[dot1+1:], '.')
+		maybeTag := strings.TrimSuffix(leaf[:dot1]+leaf[dot1+dot2+1:], ".scope")
+		if naming.ValidateSecurityTag(maybeTag) == nil {
+			return maybeTag
+		}
+	}
+	return ""
+}
+
+// PidsOfSnap returns the association of security tags to PIDs.
+//
+// NOTE: This function returns a reliable result only if the refresh-app-awareness
+// feature was enabled since all processes related to the given snap were started.
+// If the feature wasn't always enabled then only service process are correctly
+// accounted for.
+//
+// The return value is a snapshot of the pids for a given snap, grouped by
+// security tag. The result may be immediately stale as processes fork and
+// exit.
+//
+// Importantly, if the per-snap lock is held while computing the set, then the
+// following guarantee is true: if a security tag is not among the results then
+// no such tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int)
+	securityTagPrefix := "snap." + snapInstanceName + "."
+
+	// Walk the cgroup tree and look for "cgroup.procs" files. Having found one
+	// we try to derive the snap security tag from it. If successful and the
+	// tag matches the snap we are interested in, we harvest the snapshot of
+	// PIDs that belong to the cgroup and put them into a bucket associated
+	// with the security tag.
+	walkFunc := func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			// See the documentation of path/filepath.Walk. The error we get is
+			// the error that was encountered while walking. We just surface
+			// that error quickly.
+			return err
+		}
+		if fileInfo.IsDir() {
+			// We don't care about directories.
+			return nil
+		}
+		if filepath.Base(path) != "cgroup.procs" {
+			// We are looking for "cgroup.procs" files. Those contain the set
+			// of processes that momentarily inhabit a cgroup.
+			return nil
+		}
+		// Now that we are confident that the file we're looking at is
+		// interesting, extract the security tag from the cgroup path and check
+		// if the security tag belongs to the snap we are interested in. Since
+		// not all cgroups are related to snaps it is not an error if the
+		// cgroup path does not denote a snap.
+		cgroupPath := filepath.Dir(path)
+		securityTag := securityTagFromCgroupPath(cgroupPath)
+		if securityTag == "" {
+			return nil
+		}
+		if !strings.HasPrefix(securityTag, securityTagPrefix) {
+			return nil
+		}
+		pids, err := pidsInFile(path)
+		if err != nil {
+			return err
+		}
+		pidsByTag[securityTag] = append(pidsByTag[securityTag], pids...)
+		return filepath.SkipDir
```

Comment:
this probably merits a comment about looking for a single cgroup.procs file per dir ?

Reply:
Good idea

Reply:
Done